config/src/main/java/com/alibaba/nacos/config/server/service/repository/extrnal/ExternalStoragePersistServiceImpl.java

```diff
@@ -375,41 +384,53 @@ public boolean updateConfigInfo4TagCas(ConfigInfo configInfo, String tag, String
     public void insertOrUpdateBeta(final ConfigInfo configInfo, final String betaIps, final String srcIp,
             final String srcUser, final Timestamp time, final boolean notify) {
         try {
+            int updateCount = updateConfigInfo4Beta(configInfo, betaIps, srcIp, null, time, notify);
+            if (updateCount <= 0) {
+                addConfigInfo4Beta(configInfo, betaIps, srcIp, null, time, notify);
+            }
+        } catch (EmptyResultDataAccessException ex) { // No data, insert
             addConfigInfo4Beta(configInfo, betaIps, srcIp, null, time, notify);
-        } catch (DataIntegrityViolationException ive) { // Unique constraint conflict
-            updateConfigInfo4Beta(configInfo, betaIps, srcIp, null, time, notify);
         }
     }
     
     @Override
     public boolean insertOrUpdateBetaCas(final ConfigInfo configInfo, final String betaIps, final String srcIp,
             final String srcUser, final Timestamp time, final boolean notify) {
         try {
-            addConfigInfo4Beta(configInfo, betaIps, srcIp, null, time, notify);
+            boolean updateResult = updateConfigInfo4BetaCas(configInfo, betaIps, srcIp, null, time, notify);
+            if (!updateResult) {
```

Comment:
if updateConfigInfo4BetaCas fail，it may be cas check fail ,and should not execute add

Reply:
> if updateConfigInfo4BetaCas fail，it may be cas check fail ,and should not execute add

You're right, I haven't used cas, it's because I didn't consider it comprehensively